File: fixtures/scale/dart/large/lib/ciphers/caesar/caesar_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'caesar_key.dart';

/// Classic shift cipher; the key is the fixed shift amount.
class CaesarCipher implements Cipher {
  final CaesarKey key;

  CaesarCipher(this.key);

  @override
  String name() => 'caesar';

  @override
  String encrypt(String plaintext) => _shiftBy(plaintext, key.shift);

  @override
  String decrypt(String ciphertext) => _shiftBy(ciphertext, -(key.shift));

  String _shiftBy(String text, int amount) {
    final sb = StringBuffer();
    for (final c in cleanAlphabet(text).split('')) {
      sb.write(charAt(indexOf(c) + amount));
    }
    return sb.toString();
  }
}

File: fixtures/scale/scala/large/src/cryptobench/ciphers/caesar/CaesarCipher.scala
package cryptobench.ciphers.caesar

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Classic shift cipher; the key is the fixed shift amount. */
final class CaesarCipher(key: CaesarKey) extends Cipher {

  override def name: String = "caesar"

  override def encrypt(plaintext: String): String = shiftBy(plaintext, key.shift)

  override def decrypt(ciphertext: String): String = shiftBy(ciphertext, -(key.shift))

  private def shiftBy(text: String, amount: Int): String = {
    val sb = new StringBuilder
    for (c <- Alphabet.clean(text)) {
      sb.append(Alphabet.charAt(Alphabet.indexOf(c) + amount))
    }
    sb.toString
  }
}
